File: src/banking/pages/login_page.py
"""Login page object for XYZ Banking Project."""
from playwright.sync_api import Page

from src.banking.helpers.browser_helpers import (
    navigate_to_bank,
    wait_for_angular,
    _micro_delay_for_ci,
)
from src.banking.locators.selectors import (
    CustomerLoginPage,
    HomePage,
)


class LoginPage:
    def __init__(self, page: Page):
        self.page = page

    def go_to_home(self) -> None:
        """Navigate to home + wait for heading."""
        navigate_to_bank(self.page)
        self.page.wait_for_selector(HomePage.heading, state="visible", timeout=10000)

    def select_customer_login(self) -> None:
        """Click customer btn, expect user select visible."""
        self.page.locator(HomePage.customer_login_btn).click()
        wait_for_angular(self.page)
        self.page.wait_for_selector(CustomerLoginPage.user_select, state="visible", timeout=15000)

    def select_manager_login(self) -> None:
        """Click manager btn, wait for angular."""
        self.page.locator(HomePage.manager_login_btn).click()
        wait_for_angular(self.page)

    def select_user(self, name: str) -> None:
        """Select user from dropdown by label, expect login btn visible."""
        self.page.locator(CustomerLoginPage.user_select).select_option(label=name)
        self.page.wait_for_selector(CustomerLoginPage.login_btn, state="visible", timeout=5000)

    def click_login(self) -> None:
        """Click login btn, wait for angular."""
        self.page.locator(CustomerLoginPage.login_btn).click()
        wait_for_angular(self.page)
        _micro_delay_for_ci(0.15)  # CI/CD: login triggers navigation; runners need extra buffer

    def login_as_customer(self, name: str) -> None:
        """Full flow: go_to_home → select_customer_login → select_user → click_login."""
        self.go_to_home()
        self.select_customer_login()
        self.select_user(name)
        self.click_login()

File: playwright-ui/pages/LoginPage.ts
/**
 * Login page object for XYZ Banking Project.
 * Mirrors src/banking/pages/login_page.py.
 */
import type { Page } from '@playwright/test';
import { applyCiBuffer, waitForAngular } from './helpers';
import { CustomerLoginPage, HomePage } from '../locators/selectors';

/** Timeout for user select after "Customer Login" click; Angular may be slow to render. */
const USER_SELECT_TIMEOUT = 25_000;

export class LoginPage {
  constructor(private readonly page: Page) {}

  async goToHome(): Promise<void> {
    await this.page.goto('', { waitUntil: 'domcontentloaded', timeout: 30000 });
    await this.page.waitForSelector(HomePage.customer_login_btn, { state: 'visible', timeout: 15000 });
  }

  async selectCustomerLogin(): Promise<void> {
    await this.page.locator(HomePage.customer_login_btn).click();
    await waitForAngular(this.page);
    await this.page.waitForSelector(CustomerLoginPage.user_select, { state: 'visible', timeout: USER_SELECT_TIMEOUT });
  }

  async selectManagerLogin(): Promise<void> {
    await this.page.locator(HomePage.manager_login_btn).click();
    await waitForAngular(this.page);
  }

  async selectUser(name: string): Promise<void> {
    await this.page.locator(CustomerLoginPage.user_select).selectOption({ label: name });
    await this.page.waitForSelector(CustomerLoginPage.login_btn, { state: 'visible', timeout: 5000 });
  }

  async clickLogin(): Promise<void> {
    await this.page.locator(CustomerLoginPage.login_btn).click();
    await waitForAngular(this.page);
    await applyCiBuffer(0.25);
  }

  /** Full flow: goToHome → selectCustomerLogin → selectUser → clickLogin. Retries once on timeout. */
  async loginAsCustomer(name: string): Promise<void> {
    const maxAttempts = 2;
    let lastError: Error | undefined;
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        await this.goToHome();
        await this.selectCustomerLogin();
        await this.selectUser(name);
        await this.clickLogin();
        return;
      } catch (e) {
        lastError = e instanceof Error ? e : new Error(String(e));
        if (attempt < maxAttempts) {
          await new Promise((r) => setTimeout(r, 2000));
        }
      }
    }
    throw lastError;
  }
}
